feat(ECO-2863): Implement arena processing

CRBl69 opened the pull request:
# Feature progress

- [x] merge ECO-2864
- [x] merge ECO-2866
- [x] merge ECO-2867
- [x] merge ECO-2868
- [x] merge ECO-2869
- [x] merge ECO-2865
- [x] merge ECO-2909

Final implementation, largest notable changes:

- All insertions are now atomic to remove the possibility of a backfill or a processor crash resulting in partially updated state where the processor skips versions that were reported as "recorded" But had not actually successfully been inserted.

That is, the insertion logic for a transaction block now requires that the entire atomic transaction to the db succeeds before reporting it as the processor's latest transaction version, now tracked in `emojicoin_last_processed_transaction`.

- Arena candlesticks are now tracked, with an update for every new unique price ratio for two markets in an arena melee

- This processing model does light aggregation on candlestick data and "current" views of user positions and leaderboard information.

- The processing logic in `process_transactions` has been extracted into its own function, making it more modular and easy to update in the future.

- If it turns out to be non-performant or cause issues, @CRBl69 and I decided that the best course of action would be essentially splitting up the processor logic into two processors:
  - The arena processor
  - The base emojicoin processor
  - Since the logic for processing and insertion has been cleanly separated, it'd be fairly straightforward to move to a two processor model, where the previous iteration (at `4.0.2`) could exist as its own processor, and the new arena processsing code would simply be placed in a new processor, without the base `emojicoin` insertions included.

CRBl69 committed:
feat(ECO-2864): Consolidate arena migrations in one migration

CRBl69 committed:
feat(ECO-2865): Update models to work with new schema (#87)

CRBl69 committed:
feat(ECO-2866): Add arena insertion logic (#88)

CRBl69 committed:
feat(ECO-2867): Add arena leaderboard history logic (#89)

CRBl69 committed:
feat(ECO-2868): Marke arena env var optional (#90)

* feat(ECO-2868): Marke arena env var optional

* chore: clippy auto-linting `.map_or` to `.is_some_and`

* chore: Formatting

---------

Co-authored-by: Matt <90358481+xbtmatt@users.noreply.github.com>

CRBl69 committed:
feat(ECO-2869): Add candlestick processing (#93)

* feat(ECO-2869): Add candlestick processing

* chore(ECO-2909): Use new stateful processor to simplify arena logic (#94)

* chore(ECO-2909): Use new stateful processor to simplify arena logic

* feat(ECO-2923): use curve price instead of average price (#97)

* feat(ECO-2923): use curve price instead of average price

* chore(ECO-2924): final processor polishing (#96)

* chore(ECO-2924): final processor polishing

* chore(ECO-2924): improve error reporting

* chore(ECO-2924): fix arena after melee end error

xbtmatt commented on rust/processor/src/db/common/models/emojicoin_models/models/arena_candlestick.rs:
```diff
@@ -0,0 +1,136 @@
+
+        for period in periods {
+            let start_time = transaction_timestamp
+                .duration_round(period.to_time_delta())
+                .unwrap();
```
I think this function rounds up- that's not what we want here, right?

I assume you're getting the period boundary for the start time, which would always be rounded down

https://github.com/chronotope/chrono/blob/2115206bc3b0c520fe6bcc041aa724f3ad5286ea/src/round.rs#L522

xbtmatt replied:
Confirmed, just wanted to check

```rust
#[test]
fn my_test() {
    let dt = Utc
        .from_local_datetime(
            &NaiveDate::from_ymd_opt(2012, 12, 12)
                .unwrap()
                .and_hms_milli_opt(18, 24, 59, 999)
                .unwrap(),
        )
        .unwrap();
    assert_eq!(
        dt.duration_trunc(TimeDelta::try_minutes(5).unwrap())
            .unwrap()
            .to_string(),
        "2012-12-12 18:20:00 UTC"
    );
    assert_eq!(
        dt.duration_round(TimeDelta::try_minutes(5).unwrap())
            .unwrap()
            .to_string(),
        "2012-12-12 18:25:00 UTC"
    );
    let dt2 = Utc
        .from_local_datetime(
            &NaiveDate::from_ymd_opt(2012, 12, 12)
                .unwrap()
                .and_hms_milli_opt(18, 22, 30, 500)
                .unwrap(),
        )
        .unwrap();
    assert_eq!(
        dt2.duration_round(TimeDelta::try_minutes(5).unwrap())
            .unwrap()
            .to_string(),
        "2012-12-12 18:25:00 UTC"
    );
}
```

Addressed in #99

xbtmatt commented on rust/processor/src/db/common/models/emojicoin_models/queries/insertion_queries.rs:
```diff
@@ -296,3 +422,43 @@ pub fn insert_arena_vault_balance_update_events_query(
+    (
+        diesel::insert_into(schema::arena_candlestick::table)
+            .values(items_to_insert)
+            .on_conflict((melee_id, period, start_time))
+            .do_update()
+            .set((
```
If we do a `ON CONFLICT DO UPDATE` here, we'd need the processor to know whether or not this data has already been processed

Otherwise any time the processor restarts in any capacity, all candlestick data will be added to the final result again

I don't think you check for this in the main loop body, do you?

CRBl69 replied:
I did not think of this. Good catch. I'm working on a solution

CRBl69 replied:
Ok actually, it seems that your latest PR fixes this.

xbtmatt commented on rust/processor/src/db/common/models/emojicoin_models/queries/leaderboard.sql:
```diff
@@ -0,0 +1,113 @@
+-- Insert a snapshot of the leaderboard in arena_leaderboard_history for the
+-- given melee ID.
+
+-- The performance of this query has not been established, but should be
+-- irrelevant, as arenas end once every 24 hours, and this script only runs
+-- when an arena ends
```
```suggestion
-- This query hasn't been optimized, but as long as it isn't egregiously
-- inefficient it isn't a significant concern, since it's only used once
-- every 24 hours when an arena ends.
```

xbtmatt replied:
Addressed in #99

xbtmatt commented on rust/processor/src/db/common/models/emojicoin_models/queries/leaderboard.sql:
```diff
@@ -0,0 +1,113 @@
+SELECT
+    position."user",
+    $1 AS melee_id,
+    -- Proifts = Withdrawals in APT + current emojicoin balance converted to APT.
```
```suggestion
    -- Profits = Withdrawals in APT + current emojicoin balance converted to APT.
```

xbtmatt replied:
Addressed in #99

xbtmatt commented on rust/processor/src/db/common/models/emojicoin_models/queries/leaderboard.sql:
```diff
@@ -0,0 +1,113 @@
+    deposits AS losses,
+    emojicoin_0_balance,
+    emojicoin_1_balance,
+    -- If user has no balance at the end of the melee, or he has, but there is
+    -- an exit event that happened after the end of the melee, then set exited
+    -- to true, otherwise to false.
```
```suggestion
    -- Determine whether the user has exited or not by checking that one of the
    -- following conditions is true:
    --   1. The user has no balance at the end of the melee.
    --   2. There is an exit event for the user *after* that melee ended.
```

Also, after rewriting this and understanding better what it's doing, I'm wondering if this is always strictly true...are the emojicoin balances solely coming from the processor or would transfers in and out of their account (unrelated to the melee) affect their perceived "exit" status here?

CRBl69 replied:
How could they move their escrow balance ?

Is that even possible ?

xbtmatt replied:
Nope! I didn't realize it was a MyLibertyCoinStore (Alex's favorite term for unseizable coins in a `store` resource field 😆 ). Only the contract can move anything in and out

Was just making sure, and sort of thinking out loud

xbtmatt commented on rust/processor/src/db/common/models/emojicoin_models/models/arena_candlestick.rs:
```diff
@@ -0,0 +1,136 @@
+        price_0: BigDecimal,
+        price_1: BigDecimal,
+    ) -> Vec<Self> {
+        let periods = vec![
+            Period::FifteenSeconds,
+            Period::OneMinute,
+            Period::FiveMinutes,
+            Period::FifteenMinutes,
+            Period::ThirtyMinutes,
+            Period::OneHour,
+        ];
```
How come no `FourHour` candlesticks? I can kind of understand no `OneDay`

CRBl69 replied:
Since melees last 24h, 4h candles would show 6 candles. It seemed a little weird to me, but I can add it if you think it's needed.

xbtmatt replied:
You have a good point, that might look a little silly

xbtmatt commented on rust/processor/src/db/common/models/emojicoin_models/queries/insertion_queries.rs:
```diff
@@ -233,6 +227,138 @@ pub fn insert_arena_melee_events_query(
+    sql_query(query)
+}
+
+pub fn update_arena_leaderboard_history_query(
```
```suggestion
pub fn update_user_last_exit_status_after_melee_has_ended_query(
```

Or a comment down below...or something just much more explicit- it's pretty vague like this and confusing when called in the code below

xbtmatt commented on rust/processor/src/db/common/models/emojicoin_models/queries/insertion_queries.rs:
```diff
@@ -296,3 +422,43 @@ pub fn insert_arena_vault_balance_update_events_query(
+) {
+    use schema::arena_candlestick::dsl::*;
+    (
+        diesel::insert_into(schema::arena_candlestick::table)
+            .values(items_to_insert)
+            .on_conflict((melee_id, period, start_time))
+            .do_update()
+            .set((
+                open_price.eq(sql("COALESCE(")
+                    .bind(open_price)
+                    .sql(",")
+                    .bind(excluded(open_price))
+                    .sql(")")),
+                high_price.eq(sql("GREATEST(COALESCE(")
+                    .bind(high_price)
+                    .sql(",")
+                    .bind(excluded(high_price))
+                    .sql("),")
+                    .bind(excluded(high_price))
+                    .sql(")")),
+                low_price.eq(sql("LEAST(COALESCE(")
+                    .bind(low_price)
+                    .sql(",")
+                    .bind(excluded(low_price))
+                    .sql("),")
+                    .bind(excluded(low_price))
+                    .sql(")")),
+                close_price.eq(excluded(close_price)),
+                volume.eq(volume + excluded(volume)),
+                n_swaps.eq(n_swaps + excluded(n_swaps)),
+            )),
```
Aggregating candlestick data like this seems a little fragile- how do you know you haven't already added this volume and these swaps for this candlestick? The candlestick needs to track the latest swap accounted for and only insert conditionally. so it'd be fairly read and write heavy and isn't self-correcting

I think in general though this really should just be a separate service, since it's aggregating data but also tracking live data. The implementation would be pretty simple:

1. Service for candlestick processor: process grpc stream like normal but emit (aka insert) candlesticks when a period "ends". Select the swaps in that period, get the first, last, lowest, and highest price and sum the volume. This is exactly how the contract does it. it'd use the exact data model you already made. tbh this could probably just go in the current processor if it's simple enough
3. Service for the current candlestick (could possibly just be added to the broker, since it already sort of does half of this):
  - Load all 7-8 current candlesticks (one for each period) in memory by querying a view for all swap data since the last period boundary. This might be a slightly expensive read if there's huge activity, but it occurs a single time upon instantiation
  - Update the in-memory candlesticks as data comes in
  - Emit all candlestick updates to the client with websockets (already does this, just with way more data than is necessary)

xbtmatt replied:
On another note, we can verify the validity of this candlestick logic by adding extra event emission logic in e2e tests to track the custom candlesticks in the contract directly. then just compare results at the end

No need to manually calculate things like volume, highs, lows, etc in tests, just set up the test cases properly and check the final state against the contract

xbtmatt replied:
Another thing to consider is that each candlestick will almost immediately be incorrect/outdated once it's queried anyway, especially if there's a lot of activity

xbtmatt replied:
e: I jsut remembered it's more complicated because we wanted to get really accurate ratio candlesticks, not just normal ones

CRBl69 replied:
Well, your PR seems to fix this.

CRBl69 replied:
I do not see why you want to separate this into other services.

The last version thing is valid, but the same exact error could have happened in another service.

It is addressed now anyway in your newest PR.

Let's keep it this way.

CRBl69 replied:
Live candlesticks are already present in the database as the latest candlestick for a melee for a period.

I think during the broker update design call, we said that the frontend would get the initial candlestick data, then update it using the events it receives.

This logic could be moved somewhere else, but not in the broker. The scope of the broker is clearly defined, and no data processing should happen in there.

xbtmatt replied:
Not sure if you misunderstood but the data processing is super light, I'm just referring to a really simple in-memory store of recent candlestick data. It would go straight to each client. Afaik this is pretty common for candlestick data with pubsub models, and is totally in scope for the broker, since half of its purpose is just serving new candlestick data

Also how are you gonna add a whole new processor laid over the original processor and then say 30 lines added to the broker is out of scope 🤣

It'd be super simple and we wouldn't have to write to candlesticks until they're completed. It doesn't really matter now, especially since we're only tracking arena candlesticks, but I think it would be worth exploring in new iterations of these services

xbtmatt commented on rust/processor/src/processors/emojicoin_dot_fun/processor.rs:
```diff
@@ -135,164 +183,696 @@ async fn insert_to_db(
+                                        let candlestick =
+                                            ArenaCandlestickDiffModelBuilder::from_state_event(
+                                                melee_data.melee_id.clone(),
+                                                // We know at this point that the swap is in swaps
+                                                // because the Swap event is emitted before its
+                                                // corresponding State event.
+                                                // We call rev as is should be the latest swap.
```
Outdated comment? Seems like it was for the previous swap event stuff

xbtmatt replied:
Addressed in #99

xbtmatt commented on rust/processor/src/processors/emojicoin_dot_fun/processor.rs:
```diff
@@ -135,164 +183,696 @@ async fn insert_to_db(
+                                    insert_events
+                                        .arena_leaderboard_history
+                                        .push(ArenaLeaderboardHistoryPartialModel::new(melee_data));
+                                }
```
how is the first leaderboard inserted then?

CRBl69 replied:
When the first melee event is emitted, there is no previous melee, so there is nothing to insert into the leaderboard history.

xbtmatt replied:
Added as a comment

xbtmatt commented on rust/processor/src/processors/emojicoin_dot_fun/processor.rs:
```diff
@@ -135,164 +183,696 @@ async fn insert_to_db(
+                                let model = ArenaExitEventModel::new(
+                                    txn_info.clone(),
+                                    exit,
+                                    // This can never be None.
+                                    melee_data.as_ref().unwrap(),
```
```suggestion
                                    melee_data.as_ref().expect("Exit event should always appear after melee data is loaded in state."),
```

xbtmatt replied:
Addressed in #99

xbtmatt commented on rust/processor/src/processors/emojicoin_dot_fun/processor.rs:
```diff
@@ -135,164 +183,696 @@ async fn insert_to_db(
+                                insert_events.arena_exit_events.push(model)
+                            },
+                            ArenaEvent::Swap(swap) => {
+                                ensure!(
+                                    last_two_states.0.is_some() && last_two_states.1.is_some(),
+                                    "The two previous State events related to this ArenaSwap event were not found."
+                                );
+                                let (mut state_0, mut state_1) = (None, None);
+                                std::mem::swap(&mut state_0, &mut last_two_states.0);
+                                std::mem::swap(&mut state_1, &mut last_two_states.1);
+                                let (state_0, state_1) = (state_0.unwrap(), state_1.unwrap());
+                                let (state_0, state_1) = if swap.emojicoin_0_proceeds.is_zero() {
+                                    (state_1, state_0)
+                                } else {
+                                    (state_0, state_1)
+                                };
```
```suggestion
                                let state_a = last_two_states.0
                                    .take()
                                    .expect("The first state for this arena swap should be processed already.");
                                let state_b = last_two_states.1
                                    .take()
                                    .expect("The second state for this arena swap should be processed already.");

                                // Ensure the two state events are in the correct order.
                                let (state_0, state_1) = if swap.emojicoin_0_proceeds.is_zero() {
                                    (state_b, state_a)
                                } else {
                                    (state_a, state_b)
                                };
```

I believe this is more idiomatic?

Also, I'm a little confused about why emojicoin 0 proceeds means it's the second state event, and not the first..? What's the reasoning behind that? Just trying to understand

xbtmatt reviewed (commented):
Submitting this review for now, basically my initial pass at all the final changes. The last thing I really need to look over still is `leaderboard.sql`

And the candlestick update logic might need some checks to ensure you're not adding data to the same candlestick twice (or more)

---

Side note: do you mind changing `_: Option<u64>` to `_db_chain_id: Option<u64>`? Just to be clear what we're ignoring in `process_transactions(...)`

It won't let me comment on the diff...

xbtmatt commented on rust/processor/src/db/postgres/migrations/2025-01-08-171729_emojicoin_arena/up.sql:
```diff
@@ -125,17 +137,38 @@ CREATE TABLE arena_info (
+    period period_type NOT NULL,
+    start_time TIMESTAMP NOT NULL,
+
+    open_price NUMERIC,
+    high_price NUMERIC,
+    low_price NUMERIC,
+    close_price NUMERIC,
+    volume NUMERIC NOT NULL,
+    n_swaps NUMERIC NOT NULL,
```
```suggestion
    open_price NUMERIC,
    high_price NUMERIC,
    low_price NUMERIC,
    close_price NUMERIC,
    volume NUMERIC NOT NULL,
    n_swaps NUMERIC NOT NULL,
    last_version BIGINT NOT NULL,
```

The latest candlestick needs to have enough information for the client for it to be able to build off of it. `n_swaps` doesn't really indicate where in time it stopped recording (for the current candlestick), but I think `last_version` is just way simpler, it's more obvious

xbtmatt reviewed (commented):
1. Pnl seems off, but also it's partly just because it's calculated sort of weird

It seems to think I've made a profit as soon as I've entered? Which doesn't make sense. You should only calculate realized profit- unrealized profit doesn't really make sense, especially given how illiquid the markets are

2. The candlesticks are incorrect- basically, the candlestick has a `melee_id` of 30, but the volume for it is for a swap event from `melee_id: 29`

Candlestick, where data for `melee_id: 30` is showing data from `melee_id: 29`:

```json
{
  "melee_id": 30,
  "period": "period_1m",
  "start_time": "2025-03-02T06:52:00",
  "open_price": 6088.866855821374,
  "high_price": 6088.866855821374,
  "low_price": 6088.866855821374,
  "close_price": 6088.866855821374,
  "volume": 955865,
  "n_swaps": 1
}
```

Whereas the melee ids are:

```json5
{
  "melee_id": 30,
  "volume": 0, // You can see it doesn't even have volume actually
  "emojicoin_0_symbols": [ "🍧" ],
  "emojicoin_1_symbols": [ "🎟️" ],
  "emojicoin_0_market_id": 5,
  "emojicoin_1_market_id": 9,
},
{
  "melee_id": 29,
  "volume": 4852228,
  "emojicoin_0_symbols": [ "🌨️" ],
  "emojicoin_1_symbols": [ "🎟️" ],
  "emojicoin_0_market_id": 7,
  "emojicoin_1_market_id": 9,
}
```

FWIW: I've recorded the relevant parts of the JSON response for the txn version at `7777`, where it recorded it incorrectly. It's because the melee creation occurs BEFORE an exit/swap event, when the user cranks the melee due to an exit...-.- 

```json
{
  "version": "7777",
  "hash": "0x673a93b6ec719fef5e7b62747f1859f6ff004f6a4975b334989f4d9a466b2190",
  "state_change_hash": "0xd612cbb9907480c8029b3496df0e536adb596d80cc992d1b911408584b642750",
  "event_root_hash": "0x9dfe427eb7cfa7249841747863d51f7a79a3d76fae48d6709bba97d7e50271b4",
  "state_checkpoint_hash": null,
  "gas_used": "952",
  "success": true,
  "vm_status": "Executed successfully",
  "accumulator_root_hash": "0xef62d0f373a70fdca734c3fb01d3a30e3f026de26eb90823a31f7773c86307b8",
  "events": [
    { "RANDOMNESS EVENT": ["there's like 50 of them, removed em to make it easier to read"] },
    {
      "type": "0xf000d910b99722d201c6cf88eb7d1112b43475b9765b118f289b5d65d919000d::emojicoin_arena::Melee",
      "data": {
        "available_rewards": "100000000000",
        "duration": "60000000",
        "emojicoin_0_market_address": "0xdd18dfc06dd02ff2df60e0cd5d9495561cd4d3718c883e259ab14b033859b440",
        "emojicoin_1_market_address": "0x106119a0482bc7d09196bc37008a0d640aea8540650354794e583b160bc12f6f",
        "max_match_amount": "500000000",
        "max_match_percentage": "50",
        "melee_id": "30",
        "start_time": "1740898335501934"
      }
    },
    {
      "type": "0xf000d910b99722d201c6cf88eb7d1112b43475b9765b118f289b5d65d919000d::emojicoin_dot_fun::Swap",
      "data": {
        "avg_execution_price_q64": "149836233400508",
        "balance_as_fraction_of_circulating_supply_after_q64": "0",
        "balance_as_fraction_of_circulating_supply_before_q64": "18446744073709551616",
        "base_volume": "117679126229",
        "input_amount": "117679126229",
        "integrator": "0xf000d910b99722d201c6cf88eb7d1112b43475b9765b118f289b5d65d919000d",
        "integrator_fee": "4803",
        "integrator_fee_rate_bps": 50,
        "is_sell": true,
        "market_id": "9",
        "market_nonce": "5",
        "net_proceeds": "955865",
        "pool_fee": "0",
        "quote_volume": "955865",
        "results_in_state_transition": false,
        "starts_in_bonding_curve": true,
        "swapper": "0xffc117086980d34dc3b5a42cb407ed888f60623f46021f35c2ca522ea13cc961",
        "time": "1740898344410767"
      }
    },
    {
      "type": "0xf000d910b99722d201c6cf88eb7d1112b43475b9765b118f289b5d65d919000d::emojicoin_dot_fun::State",
      "data": {
        "clamm_virtual_reserves": {
          "base": "4900000000000000",
          "quote": "40000000002"
        },
        "cpamm_real_reserves": {
          "base": "0",
          "quote": "0"
        },
        "cumulative_stats": {
          "base_volume": "475482993650",
          "integrator_fees": "100019455",
          "n_chat_messages": "0",
          "n_swaps": "4",
          "pool_fees_base": "0",
          "pool_fees_quote": "0",
          "quote_volume": "3871883"
        },
        "instantaneous_stats": {
          "fully_diluted_value": "36734693879",
          "market_cap": "0",
          "total_quote_locked": "2",
          "total_value_locked": "36734693881"
        },
        "last_swap": {
          "avg_execution_price_q64": "149836233400508",
          "base_volume": "117679126229",
          "is_sell": true,
          "nonce": "5",
          "quote_volume": "955865",
          "time": "1740898344410767"
        },
        "lp_coin_supply": "0",
        "market_metadata": {
          "emoji_bytes": "0xf09f8e9fefb88f",
          "market_address": "0x106119a0482bc7d09196bc37008a0d640aea8540650354794e583b160bc12f6f",
          "market_id": "9"
        },
        "state_metadata": {
          "bump_time": "1740898344410767",
          "market_nonce": "5",
          "trigger": 3
        }
      }
    },
    {
      "type": "0xf000d910b99722d201c6cf88eb7d1112b43475b9765b118f289b5d65d919000d::emojicoin_dot_fun::Swap",
      "data": {
        "avg_execution_price_q64": "150589246413350",
        "balance_as_fraction_of_circulating_supply_after_q64": "18446744073709551616",
        "balance_as_fraction_of_circulating_supply_before_q64": "0",
        "base_volume": "116505264831",
        "input_amount": "955865",
        "integrator": "0xf000d910b99722d201c6cf88eb7d1112b43475b9765b118f289b5d65d919000d",
        "integrator_fee": "4779",
        "integrator_fee_rate_bps": 50,
        "is_sell": false,
        "market_id": "7",
        "market_nonce": "6",
        "net_proceeds": "116505264831",
        "pool_fee": "0",
        "quote_volume": "951086",
        "results_in_state_transition": false,
        "starts_in_bonding_curve": true,
        "swapper": "0xffc117086980d34dc3b5a42cb407ed888f60623f46021f35c2ca522ea13cc961",
        "time": "1740898344410767"
      }
    },
    {
      "type": "0xf000d910b99722d201c6cf88eb7d1112b43475b9765b118f289b5d65d919000d::emojicoin_dot_fun::State",
      "data": {
        "clamm_virtual_reserves": {
          "base": "4899883494735169",
          "quote": "40000951088"
        },
        "cpamm_real_reserves": {
          "base": "0",
          "quote": "0"
        },
        "cumulative_stats": {
          "base_volume": "596778754851",
          "integrator_fees": "100029455",
          "n_chat_messages": "0",
          "n_swaps": "5",
          "pool_fees_base": "0",
          "pool_fees_quote": "0",
          "quote_volume": "4861980"
        },
        "instantaneous_stats": {
          "fully_diluted_value": "36736440792",
          "market_cap": "951108",
          "total_quote_locked": "951088",
          "total_value_locked": "36736440772"
        },
        "last_swap": {
          "avg_execution_price_q64": "150589246413350",
          "base_volume": "116505264831",
          "is_sell": false,
          "nonce": "6",
          "quote_volume": "951086",
          "time": "1740898344410767"
        },
        "lp_coin_supply": "0",
        "market_metadata": {
          "emoji_bytes": "0xf09f8ca8efb88f",
          "market_address": "0x2cb9260ab4ef0264d1c871449a419e859240ba034d9f84178467794744615af5",
          "market_id": "7"
        },
        "state_metadata": {
          "bump_time": "1740898344410767",
          "market_nonce": "6",
          "trigger": 2
        }
      }
    },
    {
      "type": "0xf000d910b99722d201c6cf88eb7d1112b43475b9765b118f289b5d65d919000d::emojicoin_arena::Swap",
      "data": {
        "emojicoin_0_exchange_rate": {
          "base": "4899883494735169",
          "quote": "40000951088"
        },
        "emojicoin_0_proceeds": "116505264831",
        "emojicoin_1_exchange_rate": {
          "base": "4900000000000000",
          "quote": "40000000002"
        },
        "emojicoin_1_proceeds": "0",
        "integrator_fee": "9582",
        "melee_id": "29",
        "quote_volume": "951086",
        "user": "0xffc117086980d34dc3b5a42cb407ed888f60623f46021f35c2ca522ea13cc961"
      }
    },
    {
      "type": "0xf000d910b99722d201c6cf88eb7d1112b43475b9765b118f289b5d65d919000d::emojicoin_arena::Exit",
      "data": {
        "emojicoin_0_exchange_rate": {
          "base": "4899883494735169",
          "quote": "40000951088"
        },
        "emojicoin_0_proceeds": "116505264831",
        "emojicoin_1_exchange_rate": {
          "base": "4900000000000000",
          "quote": "40000000002"
        },
        "emojicoin_1_proceeds": "0",
        "melee_id": "29",
        "tap_out_fee": "0",
        "user": "0xffc117086980d34dc3b5a42cb407ed888f60623f46021f35c2ca522ea13cc961"
      }
    }
  ]
}
```

3. Candlesticks' open values should be the previous candlestick's close- I don't think you ever look that up anywhere (could be wrong though)

CRBl69 commented:
Addressing a couple of your points here as there are some messages I cannot directly respond to (thanks github).

> 1. Pnl seems off, but also it's partly just because it's calculated sort of weird
>
> It seems to think I've made a profit as soon as I've entered? Which doesn't make sense. You should only calculate realized profit- unrealized profit doesn't really make sense, especially given how illiquid the markets are

I think unrealized PnL is very important. Every platform I ever used (TBH I didn't use that many BUT STILL) shows unrealized PnL. Users will want to know that.

It is normal that as soon as you enter your PnL goes up. You make the price go up.


> 2. The candlesticks are incorrect- basically, the candlestick has a melee_id of 30, but the volume for it is for a swap event from melee_id: 29

I think this is very confusing, due to the nature of Frankenstein candlesticks, but here is what is happening:

- Melee 29 ends due to a melee event being emitted. This melee ENDED. There will be NO MORE candlesticks for it.
- Melee 30 starts due to the melee event.
- A swap is happening on melee 29, because you can swap after the melee ended.
- For the swap to happen, 🎟️ needs to be traded.
- 🎟️ part of both melee 29 AND 30.
- 🎟️ is traded during melee 30.
- Price of 🎟️ changes during melee 30.
- The price change of 🎟️, the volume and other data of 🎟️ needs to be reflected in the candlestick for melee 30.

IMO, the logic is correct here.

> 3. Candlesticks' open values should be the previous candlestick's close- I don't think you ever look that up anywhere (could be wrong though)

We have discussed this on ENG call with you and Alex and said that the open price is the price of at which the first swap happened. However, this was before switching to curve price. Do you think it does not make sense to use this logic with curve price ? (I could see why)

xbtmatt committed:
fix(ECO-2939): Arena processing tweaks/fixes (#99)

* fix(ECO-2939): Arena processing tweaks/fixes

* chore: Reword/fix comments in leaderboard.sql

* chore: Remove outdated comment

* chore: Use more idiomatic `expect(...)` and `take(...).expect(...)` instead of `unwrap` and `std::mem::swap`

* chore: Add meaningful names to the two state events, since the order actually matters

* chore(ECO-2940): Add `last_transaction_version` for tables that track the latest state of something (#100)

* Add `last_transaction_version` for tables that track the latest state of something

* fix(ECO-2945): Only process unprocessed transactions (#101)

* fix(ECO-2945): Only process unprocessed transactions

* chore(ECO-2946): Add `last_transaction_version` to all remaining tables; emit candlesticks; truncate candlestick prices (#102)

* chore(ECO-2946): Add `last_transaction_version` to all remaining tables

* fix: Add the last transaction version to the `leaderboard.sql` query properly

* Change `ArenaCandlestickDiffModel` => `ArenaCandlestickModel` because they are the full models at that point

Add them to `EmojicoinDbEvent` type so the broker can use them

* Fix name

* Round candlestick decimals to 16 decimal places

* Rename `arena_candlestick` to `arena_candlesticks`

* Satisfy clippy

---------

Co-authored-by: Matt <90358481+xbtmatt@users.noreply.github.com>

---------

Co-authored-by: Bogdan Crisan <bogdan@crbl.studio>

---------

Co-authored-by: Bogdan Crisan <bogdan@crbl.studio>

xbtmatt committed:
Clarify comments

xbtmatt commented:
You actually have a good point about people wanting to see unrealized, but my main point was that the only thing that makes sense for the leaderboard is realized pnl- otherwise it's *always* beneficial for someone to never close their position before the end of a melee, because "slippage" would make their pnl worse (and in some cases with large positions, it could mean the difference between +50% and -20%)

Something we can revisit later

xbtmatt reviewed (approved):
The end of an era! Thanks for so many iterations on this. Excited to use it!!

xbtmatt committed:
Make candlesticks non nullable

xbtmatt committed:
Make candlesticks non nullable in schema.rs

CRBl69 committed:
fix(ECO-2863): use correct version in processor state
